Fix and regenerate invalid block fixtures

aduth opened the pull request:
Related: https://github.com/WordPress/gutenberg/pull/1929#issuecomment-318134153

This pull request seeks to resolve markup errors present in block fixtures to ensure that all fixtures are valid. As described in https://github.com/WordPress/gutenberg/pull/1929#issuecomment-318134153, we'll want some intentionally-invalid examples as well, but those are not included in the changes here.

The changes also add some nice conveniences to better compare a failed parse between its actual and expected representation. For example:

![Parse fail](https://user-images.githubusercontent.com/1779930/28682528-4e7e5dc2-72cb-11e7-82c2-a653300dd7c1.png)

This works only in Node environments currently, due to differences in log coloring between Node and browsers. This is something I'd like to consider improving, however separately.

There were a number of failed tests in `api/test/parser.js` after introducing these changes that I am surprised were ever working, since the dummy block implementations were clearly flawed. Those have been resolved here.

__Testing instructions:__

Verify that tests pass when regenerating fixtures from scratch:

```
./node_modules/.bin/rimraf blocks/test/fixtures/*.serialized.html && \
    ./node_modules/.bin/rimraf blocks/test/fixtures/*.json && \
    GENERATE_MISSING_FIXTURES=y npm run test-unit
```

Ensure there are no regressions with revised pullquote and gallery attribute defaults (moved to default attributes for consistency between `edit` and `save` usage).

aduth committed:
Log parse errors in all environments except production

aduth committed:
Improve error messaging for parse failures

aduth committed:
Fix and regenerate invalid fixtures

nylen commented:
> we'll want some intentionally-invalid examples as well, but those are not included in the changes here

I think we might as well handle this in this PR, too: some of the fixtures modified in this PR should serve as our intentionally-invalid examples.

nylen commented on blocks/api/parser.js:
```diff
@@ -123,13 +123,54 @@ export function isValidBlock( rawContent, blockType, attributes ) {
+			// ANSI codes are supported. In these environments we provide a nicely
+			// colored error message for invalid parse.
+			//
+			// TODO: Teach browsers to ANSI
```
... or teach `chalk` to output HTML.  ANSI escape sequences are quite convoluted, and it seems like what we really want is an intermediate representation that can be outputted in either form.

aduth replied:
>... or teach `chalk` to output HTML.  ANSI escape sequences are quite convoluted, and it seems like what we really want is an intermediate representation that can be outputted in either form.

I did run across this: https://www.npmjs.com/package/ansi_up

I'd also toyed with ideas for converting ANSI automatically to [CSS-styled `console.log`](https://stackoverflow.com/a/13017382) depending on the environment.

nylen commented on blocks/api/parser.js:
```diff
@@ -123,13 +123,54 @@ export function isValidBlock( rawContent, blockType, attributes ) {
+				// eslint-disable-next-line no-console
+				console.warn( message );
+			}
+		}
```
I think this code block could be structured better to achieve the following goals:

- Serve as a foundation for an eventual UI that explains the differences to the user.
- Specifically test the logic for generating the difference messages.

We probably do want this code in the production build eventually, so these changes would end up removing most of the environment-specific conditionals.  What would remain would be something like this:

```js
if ( 'test' === process.env.NODE_ENV ) {
    // Maybe we can even remove this, in favor of
    // `expect( console.warn ).toHaveBeenCalledWith( ... )`
    // in the tests below.
    throw new Error( message );
} else if ( 'production' !== process.env.NODE_ENV ) {
    console.warn( message );
}
```

aduth replied:
The `NODE_ENV` condition wasn't so much for testing the parser as much as it was to ensure the fixtures testing would actually log the failure if the parse was invalid (`console.warn` is not sufficient).

Instead, in cdde9c4 I've changed it so that the error is thrown in all cases. It is handled by a `try` / `catch` in `createBlockWithFallback` which assigns the error into the block. With a reintroduction of the invalid blocks, the fixtures testing determines whether the parse _should_ have failed and otherwise surfaces the error to Jest.

I went back and forth on this a bit, trying to decide if throwing an error was the right path. It's otherwise difficult to detect the failure in fixtures testing... except with overriding `console.error`, which felt a bit hacky. I think they're legitimate errors and should be treated as such.

The downside is that because the error is handled, it is no longer logged to the console in the browser. But because the error is now stored into the block, I think we have some better opportunities to surface the message through the UI, rather than through the console (even in the "Invalid Block" dialog).

To simplify the formatting, I've removed the `chalk.isEnabled` split and just use chalk in all cases. It degrades gracefully to `identity` when not enabled.

nylen commented on blocks/api/test/parser.js:
```diff
@@ -97,11 +98,16 @@ describe( 'block parser', () => {
 				{ fruit: 'Bananas' }
 			) ).toBe( false );
+
+			process.env.NODE_ENV = env;
```
I'd rather see this structured as follows:

```js
expect( () => isValidBlock( ... ) ).toThrow( ... );
```
or, better still:
```js
expect( isvalidBlock( ... ) ).toBe( false );
expect( console.warn ).toHaveBeenCalledWith( ... );
```

aduth replied:
Per above https://github.com/WordPress/gutenberg/pull/2056#discussion_r131038937, this now tests that `isValidBlock` throws.

aduth committed:
Parser: Consolidate error to common message

Chalk still works when not enabled, but doesn't add any styles. This was avoided previously because messaging looked poor on per-character diff. We later changed to per-line diff with prefix, which looks reasonable in no-style environments. Further, prefixing retains readability for console.error, so drop thrown error.

aduth committed:
Fixtures: Add and test intentionally invalid blocks

aduth commented:
>I think we might as well handle this in this PR, too: some of the fixtures modified in this PR should serve as our intentionally-invalid examples.

The previous invalid versions of the blocks were restored and tested as intentionally invalid as of cdde9c4.

nylen commented on blocks/api/test/parser.js:
```diff
@@ -97,11 +98,11 @@ describe( 'block parser', () => {
 				getBlockType( 'core/test-block' ),
 				{ fruit: 'Bananas' }
-			) ).toBe( false );
+			) ).toThrow();
```
Should we test a specific error message here, or save this for later when we're ready to expose this info to the UI?

aduth replied:
Is a pain to test with the ANSI escapes, unless we were to disable that for testing, or just test the pattern.

>when we're ready to expose this info to the UI?

Which I'm starting to wonder, do we want the error message exposed in the UI anywhere? Currently we're only generating and throwing the error in development modes, which means it's not something generally accessible in the UI (in distributed plugin anyways).

nylen replied:
To clarify, I think what should be shown in the UI is more like a text-mode diff and a prompt to confirm that the changes are acceptable.

nylen committed:
Add comments to break up full-content testing into steps

nylen commented on blocks/test/full-content.js:
```diff
@@ -165,6 +198,11 @@ describe( 'full post content fixture', () => {
+
 			// `serialize` doesn't have a trailing newline, but the fixture
 			// files should.
 			const serializedActual = serialize( blocksActual ) + '\n';
```
Aside: I thought we got rid of this at some point?

nylen commented on blocks/test/full-content.js:
```diff
@@ -152,6 +167,24 @@ describe( 'full post content fixture', () => {
+			if ( ! isExpectedValid ) {
+				return;
+			}
+
```
I think for an invalid parse result, we should still verify that the block is initialized and serialized as expected (currently, no change).

aduth commented on blocks/test/full-content.js:
```diff
@@ -95,13 +95,23 @@ describe( 'full post content fixture', () => {
 
 	fileBasenames.forEach( f => {
 		it( f, () => {
+			/**
```
We have three styles of multi-line comments:

- [WordPress standard](https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/#comments) `/*`
- [Calypso standard](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#comments) `//`
- This `/**`

nylen replied:
I think the WordPress standard was written before JSDoc came into wide usage.

I don't really have a preference here, do you?  I just wrote them that way so they would visually look like separate sections.

aduth replied:
Mine is `//`, but I also didn't realize 'til recently that this was at odds with the core convention 😬  I'd say I have a preference _against_ this style, only because it is easy to confuse with ([and encounter errors as](http://eslint.org/docs/rules/valid-jsdoc)) JSDoc.

nylen replied:
I went with `/*` to preserve the "visual separator" aspect of these comments.

nylen commented on blocks/test/full-content.js:
```diff
@@ -95,14 +95,22 @@ describe( 'full post content fixture', () => {
 				);
 			}
 
-			const blocksActual = parse( content );
```
This declaration should stay below with the rest of the "list of block objects" stuff.  I've added comments to better delineate this code into logical steps.  In a separate PR, we could try adding new `it()` test cases for each of tehse steps.

nylen committed:
Tweak comment delimiters

aduth commented:
This pull request will need some further thought, particularly around validation errors, which might ought to be split to its own pull request. Notably with #2210, we no longer use beautification to validate post content, so simply logging the "actual" and "expected" contents might not be the most useful to illustrate the true difference that was captured.

A couple considerations:

- Beautify contents in logging only
- Leverage the fact that we know the precise token where validation failed and highlight it in logged error message. Might be difficult to log in context of the full text, although location [can be added into the tokens without much trouble](https://github.com/tildeio/simple-html-tokenizer/blob/f8bd1b150cb78c47ca8d6e8e713f89dd8c8b5cd8/src/tokenizer.ts#L65-L76), at which point it might just be some simple slicing.

But as mentioned at https://github.com/WordPress/gutenberg/pull/2056#discussion_r131424413, there's also some open questions about the sorts of errors and messaging we want to be able to display _in the UI_, not just development-mode `console.error`.

I might see about piecing apart this pull request into separate ones, as the original goal (title) has certainly drifted.

nylen commented:
I'd propose that in this PR we keep the fixture file changes and undo the changes to the parser API.  I can help out with that over the next couple of days.
